app_allocator/classes/dynamic_matrix_heuristic.py

```diff
@@ -0,0 +1,176 @@
+
+        self.judge_capacities = self._calc_judge_capacities()
+        self.application_needs = self.initial_application_needs()
+
+    def assess(self):
+        pass  # pragma: nocover
+
+    def work_left(self):
+        return len(self.applications) > 0
+
+    def _calc_judge_capacities(self):
+        return {judge: int(judge['commitment']) for judge in self.judges}
+
+    def _calc_needs_matrix(self):
+        return matrix([list(row.values())
+                       for _, row in self.application_needs.items()])
+
+    def find_one_application(self, judge):
+        result = self.request_batch(judge, 1)
+        if len(result) > 0:
+            return result[0]
+        else:
+            return None
+
+    def request_batch(self, judge, batch_size):
+        if len(self.applications) <= batch_size:
+            can_assign = self._can_assign_func(judge)
+            apps = [app for app in self.applications if can_assign(app)]
+        else:
+            available_batch_size = min(batch_size,
+                                       self.judge_capacities[judge])
+            judge_features = self.judge_features(judge)
+
+            needs_array = array([list(row.values()) for _, row
+                                 in self.application_needs.items()])
+            feature_weights_array = array([v for v in
+                                           self.feature_weights.values()])
+            needs_matrix = matrix(needs_array * feature_weights_array)
+            application_preferences = judge_features * needs_matrix.transpose()
+            apps = self.choose_n_applications(judge,
+                                              application_preferences,
+                                              available_batch_size)
+        for app in apps:
+            self.judge_assignments[judge].append(app)
+            self.judge_capacities[judge] -= 1
+            self._update_needs("assigned", judge, app)
+        return apps
+
+    def choose_n_applications(self, judge, application_preferences, n):
+        applications = (array(self.applications)[
+                (-application_preferences).argsort()]).flat
+        can_assign = self._can_assign_func(judge)
+        filtered = [app for app in applications if can_assign(app)]
+        return filtered[:n]
+
+    def _can_assign_func(self, judge):
+        '''returns a function which takes an application and returns True
+        if the application can be assigned to the judge.
+        Presumably this is where we'll put the z-score logic as well.
+        '''
+        def can_assign_to_judge(application):
+            return application not in self.judge_assignments[judge]
+
+        return can_assign_to_judge
+
+    def _feature_values(self, entity_sets):
+        return tuple(set([(feature, entity[feature.field])
+                          for entities in entity_sets
+                          for entity in entities
+                          for feature in self.features]))
```

Comment:
Glad you liked it! The `set` is because this iterates over all of the existing entities (applications and judges) and extracts the (feature, value) pairs that it finds. 
This is an expensive operation, and I expect that we'll be able to improve it once we're consulting a Criterion object in the db. (for example, if the Criterion knows that it's matching on Program, we don't need to look at individual judges to know that our choices are [Boston, Israel, Switzerland, ...])
Since this is the source of truth for feature ordering, I'm not too worried about ordering getting scrambled, but I've added a sort to ensure that order is stable.